File: tests/c/hokey-pokey.c
#include <errno.h>
#include <math.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>

// Compilation is done using GCC or CLang :
// - gcc -std=c99 -Wall -Wextra -pedantic -Ofast -o hokey-pokey hokey-pokey.c -lm
// - clang -std=c99 -Wall -Wextra -pedantic -Ofast -o hokey-pokey hokey-pokey.c -lm

// This function written in C99 is not affiliated with the CIE (International Commission on Illumination),
// and is released into the public domain. It is provided "as is" without any warranty, express or implied.

// Expressly defining pi ensures that the code works on different platforms.
#ifndef M_PI
#define M_PI 3.14159265358979323846264338328
#endif

// The classic CIE ΔE2000 implementation, which operates on two L*a*b* colors, and returns their difference.
// "l" ranges from 0 to 100, while "a" and "b" are unbounded and commonly clamped to the range of -128 to 127.
static double ciede_2000(const double l_1, const double a_1, const double b_1, const double l_2, const double a_2, const double b_2) {
	// Working in C with the CIEDE2000 color-difference formula.
	// k_l, k_c, k_h are parametric factors to be adjusted according to
	// different viewing parameters such as textures, backgrounds...
	const double k_l = 1.0;
	const double k_c = 1.0;
	const double k_h = 1.0;
	double n = (hypot(a_1, b_1) + hypot(a_2, b_2)) * 0.5;
	n = n * n * n * n * n * n * n;
	// A factor involving chroma raised to the power of 7 designed to make
	// the influence of chroma on the total color difference more accurate.
	n = 1.0 + 0.5 * (1.0 - sqrt(n / (n + 6103515625.0)));
	// hypot calculates the Euclidean distance while avoiding overflow/underflow.
	const double c_1 = hypot(a_1 * n, b_1);
	const double c_2 = hypot(a_2 * n, b_2);
	// atan2 is preferred over atan because it accurately computes the angle of
	// a point (x, y) in all quadrants, handling the signs of both coordinates.
	double h_1 = atan2(b_1, a_1 * n);
	double h_2 = atan2(b_2, a_2 * n);
	if (h_1 < 0.0)
		h_1 += 2.0 * M_PI;
	if (h_2 < 0.0)
		h_2 += 2.0 * M_PI;
	n = fabs(h_2 - h_1);
	// Cross-implementation consistent rounding.
	if (M_PI - 1E-14 < n && n < M_PI + 1E-14)
		n = M_PI;
	// When the hue angles lie in different quadrants, the straightforward
	// average can produce a mean that incorrectly suggests a hue angle in
	// the wrong quadrant, the next lines handle this issue.
	double h_m = (h_1 + h_2) * 0.5;
	double h_d = (h_2 - h_1) * 0.5;
	if (M_PI < n) {
		if (0.0 < h_d)
			h_d -= M_PI;
		else
			h_d += M_PI;
		h_m += M_PI;
	}
	const double p = 36.0 * h_m - 55.0 * M_PI;
	n = (c_1 + c_2) * 0.5;
	n = n * n * n * n * n * n * n;
	// The hue rotation correction term is designed to account for the
	// non-linear behavior of hue differences in the blue region.
	const double r_t = -2.0 * sqrt(n / (n + 6103515625.0))
				* sin(M_PI / 3.0 * exp(p * p / (-25.0 * M_PI * M_PI)));
	n = (l_1 + l_2) * 0.5;
	n = (n - 50.0) * (n - 50.0);
	// Lightness.
	const double l = (l_2 - l_1) / (k_l * (1.0 + 0.015 * n / sqrt(20.0 + n)));
	// These coefficients adjust the impact of different harmonic
	// components on the hue difference calculation.
	const double t = 1.0 	+ 0.24 * sin(2.0 * h_m + M_PI / 2.0)
				+ 0.32 * sin(3.0 * h_m + 8.0 * M_PI / 15.0)
				- 0.17 * sin(h_m + M_PI / 3.0)
				- 0.20 * sin(4.0 * h_m + 3.0 * M_PI / 20.0);
	n = c_1 + c_2;
	// Hue.
	const double h = 2.0 * sqrt(c_1 * c_2) * sin(h_d) / (k_h * (1.0 + 0.0075 * n * t));
	// Chroma.
	const double c = (c_2 - c_1) / (k_c * (1.0 + 0.0225 * n));
	// Returns the square root so that the Delta E 2000 reflects the actual geometric
	// distance within the color space, which ranges from 0 to approximately 185.
	return sqrt(l * l + h * h + c * c + c * h * r_t);
}

// Compilation is done using GCC or CLang :
// - gcc -std=c99 -Wall -Wextra -pedantic -Ofast -o ciede-2000-compiled ciede-2000.c -lm
// - clang -std=c99 -Wall -Wextra -pedantic -Ofast -o ciede-2000-compiled ciede-2000.c -lm

// GitHub Project : https://github.com/michel-leonard/ciede2000-color-matching
//  More Examples : https://michel-leonard.github.io/ciede2000-color-matching/discovery-generator.html

// L1 = 44.0           a1 = -42.01         b1 = -116.09
// L2 = 44.0           a2 = -42.0          b2 = -116.09
// CIE ΔE2000 = ΔE00 = 0.00290606284

// L1 = 67.01          a1 = 25.0           b1 = 56.214
// L2 = 69.0           a2 = 20.53          b2 = 57.0
// CIE ΔE2000 = ΔE00 = 3.29500383775

// L1 = 9.4306         a1 = 69.934         b1 = -15.41
// L2 = 13.0           a2 = 69.934         b2 = -6.813
// CIE ΔE2000 = ΔE00 = 4.07350320628

// L1 = 45.364         a1 = 4.9744         b1 = -112.7
// L2 = 45.364         a2 = 13.1175        b2 = -109.0
// CIE ΔE2000 = ΔE00 = 4.70956275136

// L1 = 75.8987        a1 = 116.0          b1 = 119.0822
// L2 = 73.88          a2 = 80.0           b2 = 85.12
// CIE ΔE2000 = ΔE00 = 6.94495627335

// L1 = 81.647         a1 = 43.79          b1 = 76.66
// L2 = 79.856         a2 = 41.1478        b2 = 104.95
// CIE ΔE2000 = ΔE00 = 8.88251871921

// L1 = 54.31          a1 = -66.0          b1 = -108.701
// L2 = 44.0           a2 = -62.8882       b2 = -54.2447
// CIE ΔE2000 = ΔE00 = 15.6770738594

// L1 = 15.1158        a1 = -75.0          b1 = 3.8001
// L2 = 35.3           a2 = -122.7         b2 = -13.9
// CIE ΔE2000 = ΔE00 = 18.39394528246

// L1 = 25.805         a1 = 26.26          b1 = 45.681
// L2 = 36.314         a2 = 69.0           b2 = 37.097
// CIE ΔE2000 = ΔE00 = 23.5609617723

// L1 = 69.0           a1 = 124.04         b1 = 121.803
// L2 = 12.5171        a2 = -112.7886      b2 = 117.2
// CIE ΔE2000 = ΔE00 = 105.65854555044

/////////////////////////////////////////////////////////////////////
/////////////////////////////////////////////////////////////////////
/////////////////////////////////////////////////////////////////////
//////////////////                          /////////////////////////
//////////////////                          /////////////////////////
//////////////////         TESTING          /////////////////////////
//////////////////                          /////////////////////////
//////////////////                          /////////////////////////
/////////////////////////////////////////////////////////////////////
/////////////////////////////////////////////////////////////////////
/////////////////////////////////////////////////////////////////////

struct test_row {
	double L1;
	double a1;
	double b1;
	double L2;
	double a2;
	double b2;
	double deltaE;
};

typedef unsigned long long int u64;

static struct test_row test_row = {0};
static char buf[255] = {0};

u64 xor_random(u64 *s) {
	// A shift-register generator has a reproducible behavior across platforms.
	return *s ^= *s << 13, *s ^= *s >> 7, *s ^= *s << 17 ;
}

static double rand_double_64(double min, double max, u64 * seed) {
	// Generate a 64-bit random integer by combining 8 random bytes
	// Normalize to the range [0, 1) and scale to [min, max)
	return min + (max - min) * ((double) xor_random(seed) / 18446744073709551616.0);
}

static void prepare_values(int num) {
	u64 seed = 0x2236b69a7d223bd ^ (u64)num ^ (u64)&errno ;
	*buf = 0;
	sprintf(buf, "./values-c.txt");
	printf("prepare_values('%s', %d)\n", buf, num);
	FILE *fp = fopen(buf, "w");
	for (int i = 0; i < num; ++i) {
		struct test_row r = {
				rand_double_64(0, 100, &seed),
				rand_double_64(-128, 128, &seed),
				rand_double_64(-128, 128, &seed),
				rand_double_64(0, 100, &seed),
				rand_double_64(-128, 128, &seed),
				rand_double_64(-128, 128, &seed),
				0.
		};
		const u64 type = xor_random(&seed);
		if (type & 1) r.L1 = round(r.L1);
		if (type & 2) r.a1 = round(r.a1);
		if (type & 4) r.b1 = round(r.b1);
		if (type & 8) r.L2 = round(r.L2);
		if (type & 16) r.a2 = round(r.a2);
		if (type & 32) r.b2 = round(r.b2);
		r.deltaE = ciede_2000(r.L1, r.a1, r.b1, r.L2, r.a2, r.b2);
		fprintf(fp, "%.17g,%.17g,%.17g,%.17g,%.17g,%.17g,%.17g\n", r.L1, r.a1, r.b1, r.L2, r.a2, r.b2, r.deltaE);
		if (i % 1000 == 0) {
			putchar('.');
			fflush(stdout);
		}
	}
	fclose(fp);
}

static void compare_values(const char *ext) {
	*buf = 0;
	sprintf(buf, "./../%s/values-%s.txt", ext, ext);
	printf("compare_values('%s')\n", buf);
	FILE *fp = fopen(buf, "r");
	int n_rows = 0, n_err = 0;
	struct test_row *r = &test_row;
	while (fgets(buf, 1023, fp)) {
		++n_rows;
		buf[strlen(buf) - 1] = 0;
		char *s = strtok(buf, ",");
		int j = -1;
		while (s) {
			++j;
			if (j == 0) r->L1 = atof(s);
			else if (j == 1) r->a1 = atof(s);
			else if (j == 2) r->b1 = atof(s);
			else if (j == 3) r->L2 = atof(s);
			else if (j == 4) r->a2 = atof(s);
			else if (j == 5) r->b2 = atof(s);
			else if (j == 6) r->deltaE = atof(s);
			s = strtok(NULL, ",");
		}
		const double calc = ciede_2000(r->L1, r->a1, r->b1, r->L2, r->a2, r->b2);
		const double err = fabs(calc - r->deltaE);
		if ((calc != calc) || (1e-10 < err)) {
			printf("%d. read [%g,%g,%g] [%g,%g,%g] expect %g get %g (err=%g)\n", ++n_err, r->L1, r->a1, r->b1, r->L2,
				   r->a2, r->b2,
				   r->deltaE, calc, err);
			break;
		} else if (n_rows % 1000 == 0) {
			putchar('.');
			fflush(stdout);
		}
	}
	fclose(fp);
}

static int is_alpha(char *s) {
	for (; *s; ++s)
		if ((*s < 'a' || 'z' < *s) && (*s < 'A' || 'Z' < *s))
			return 0;
	return 1;
}

int main(int argc, char *argv[]) {
	char **end = NULL;
	if (argc && is_alpha(argv[1]))
		compare_values(argv[1]);
	else
		prepare_values(argc ? strtol(argv[1], end, 10) : 10000);
}

File: tests/php/hokey-pokey.php
<?php

// This function written in PHP is not affiliated with the CIE (International Commission on Illumination),
// and is released into the public domain. It is provided "as is" without any warranty, express or implied.

// The classic CIE ΔE2000 implementation, which operates on two L*a*b* colors, and returns their difference.
function ciede_2000($l_1, $a_1, $b_1, $l_2, $a_2, $b_2) {
	// Working in PHP with the CIEDE2000 color-difference formula.
	// k_l, k_c, k_h are parametric factors to be adjusted according to
	// different viewing parameters such as textures, backgrounds...
	$k_l = $k_c = $k_h = 1.0;
	$n = (hypot($a_1, $b_1) + hypot($a_2, $b_2)) * 0.5;
	$n = $n * $n * $n * $n * $n * $n * $n;
	// A factor involving chroma raised to the power of 7 designed to make
	// the influence of chroma on the total color difference more accurate.
	$n = 1.0 + 0.5 * (1.0 - sqrt($n / ($n + 6103515625.0)));
	// hypot calculates the Euclidean distance while avoiding overflow/underflow.
	$c_1 = hypot($a_1 * $n, $b_1);
	$c_2 = hypot($a_2 * $n, $b_2);
	// atan2 is preferred over atan because it accurately computes the angle of
	// a point (x, y) in all quadrants, handling the signs of both coordinates.
	$h_1 = atan2($b_1, $a_1 * $n);
	$h_2 = atan2($b_2, $a_2 * $n);
	$h_1 += 2.0 * M_PI * ($h_1 < 0.0);
	$h_2 += 2.0 * M_PI * ($h_2 < 0.0);
	$n = abs($h_2 - $h_1);
	// Cross-implementation consistent rounding.
	if (M_PI - 1E-14 < $n && $n < M_PI + 1E-14)
		$n = M_PI;
	// When the hue angles lie in different quadrants, the straightforward
	// average can produce a mean that incorrectly suggests a hue angle in
	// the wrong quadrant, the next lines handle this issue.
	$h_m = ($h_1 + $h_2) * 0.5;
	$h_d = ($h_2 - $h_1) * 0.5;
	if (M_PI < $n) {
		if (0.0 < $h_d)
			$h_d -= M_PI;
		else
			$h_d += M_PI;
		$h_m += M_PI;
	}
	$p = 36.0 * $h_m - 55.0 * M_PI;
	$n = ($c_1 + $c_2) * 0.5;
	$n = $n * $n * $n * $n * $n * $n * $n;
	// The hue rotation correction term is designed to account for the
	// non-linear behavior of hue differences in the blue region.
	$r_t = -2.0 * sqrt($n / ($n + 6103515625.0))
					   * sin(M_PI / 3.0 * exp($p * $p / (-25.0 * M_PI * M_PI)));
	$n = ($l_1 + $l_2) * 0.5;
	$n = ($n - 50.0) * ($n - 50.0);
	// Lightness.
	$l = ($l_2 - $l_1) / ($k_l * (1.0 + 0.015 * $n / sqrt(20.0 + $n)));
	// These coefficients adjust the impact of different harmonic
	// components on the hue difference calculation.
	$t = 1.0	+ 0.24 * sin(2.0 * $h_m + M_PI * 0.5)
			+ 0.32 * sin(3.0 * $h_m + 8.0 * M_PI / 15.0)
			- 0.17 * sin($h_m + M_PI / 3.0)
			- 0.20 * sin(4.0 * $h_m + 3.0 * M_PI / 20.0);
	$n = $c_1 + $c_2;
	// Hue.
	$h = 2.0 * sqrt($c_1 * $c_2) * sin($h_d) / ($k_h * (1.0 + 0.0075 * $n * $t));
	// Chroma.
	$c = ($c_2 - $c_1) / ($k_c * (1.0 + 0.0225 * $n));
	// Returning the square root ensures that the result represents
	// the "true" geometric distance in the color space.
	return sqrt($l * $l + $h * $h + $c * $c + $c * $h * $r_t);
}

// GitHub Project : https://github.com/michel-leonard/ciede2000-color-matching
//  More Examples : https://michel-leonard.github.io/ciede2000-color-matching/samples.html

// L1 = 27.0           a1 = 77.8536        b1 = 55.3
// L2 = 27.0           a2 = 77.855         b2 = 55.3
// CIE ΔE2000 = ΔE00 = 0.00044644116

// L1 = 88.854         a1 = 46.2267        b1 = 83.2139
// L2 = 88.9           a2 = 46.2267        b2 = 83.0
// CIE ΔE2000 = ΔE00 = 0.07202154922

// L1 = 68.24          a1 = 118.0          b1 = -98.874
// L2 = 68.24          a2 = 118.7771       b2 = -98.874
// CIE ΔE2000 = ΔE00 = 0.1649792473

// L1 = 67.7           a1 = 45.0           b1 = -24.662
// L2 = 68.02          a2 = 45.0           b2 = -24.0
// CIE ΔE2000 = ΔE00 = 0.39832644443

// L1 = 60.0           a1 = 6.64           b1 = 116.0
// L2 = 64.0           a2 = 1.0            b2 = 116.0
// CIE ΔE2000 = ΔE00 = 4.38605812736

// L1 = 51.2           a1 = -1.9839        b1 = 77.9
// L2 = 56.0           a2 = -1.9839        b2 = 77.9
// CIE ΔE2000 = ΔE00 = 4.64278946369

// L1 = 96.266         a1 = -8.1           b1 = -107.679
// L2 = 85.6554        a2 = -31.766        b2 = -120.0
// CIE ΔE2000 = ΔE00 = 11.55414446275

// L1 = 79.0           a1 = 62.0           b1 = -9.9
// L2 = 73.0           a2 = 36.6           b2 = -36.1
// CIE ΔE2000 = ΔE00 = 17.05062277347

// L1 = 9.0            a1 = 101.63         b1 = -45.0079
// L2 = 26.7           a2 = 65.2           b2 = -79.2
// CIE ΔE2000 = ΔE00 = 21.63698741748

// L1 = 73.25          a1 = 76.2           b1 = -124.7913
// L2 = 16.224         a2 = -46.31         b2 = 124.13
// CIE ΔE2000 = ΔE00 = 112.62038066456

/////////////////////////////////////////////////////////////////////
/////////////////////////////////////////////////////////////////////
/////////////////////////////////////////////////////////////////////
//////////////////                          /////////////////////////
//////////////////                          /////////////////////////
//////////////////         TESTING          /////////////////////////
//////////////////                          /////////////////////////
//////////////////                          /////////////////////////
/////////////////////////////////////////////////////////////////////
/////////////////////////////////////////////////////////////////////
/////////////////////////////////////////////////////////////////////

function prepare_values($n_lines = 10000) {
	$filename = './values-php.txt' ;
	echo "prepare_values('$filename', $n_lines)\n" ;
	$fp = fopen($filename, 'w');
	for($i = 0; $i < $n_lines; ++$i){
		$values = [
			round(100. * (mt_rand() / mt_getrandmax()), mt_rand(0, 2)),
			round(255. * (mt_rand() / mt_getrandmax()) - 128., mt_rand(0, 2)),
			round(255. * (mt_rand() / mt_getrandmax()) - 128., mt_rand(0, 2)),
			round(100. * (mt_rand() / mt_getrandmax()), mt_rand(0, 2)),
			round(255. * (mt_rand() / mt_getrandmax()) - 128., mt_rand(0, 2)),
			round(255. * (mt_rand() / mt_getrandmax()) - 128., mt_rand(0, 2)),
		];
		if ($i % 1000 == 0)
			echo '.' ;
		$values[ ] = ciede_2000(...$values);
		fputcsv($fp, $values);
	}
	fclose($fp);
}

function compare_values($extension = 'php'){
	$i = $n_errors = 0 ;
	$filename = "./../$extension/values-$extension.txt" ;
	echo "compare_values('$filename')\n" ;
	$fp = fopen($filename, 'r');
	while($arr = fgetcsv($fp, 1024, ',')){
		++$i ;
		$arr = array_map('floatval', $arr);
		$delta_e = array_pop($arr) ;
		$res = ciede_2000(...$arr) ;
		$abs_err = abs($delta_e - $res);
		if (!is_finite($delta_e) || !is_finite($res) || 1e-10 < $abs_err) {
			echo json_encode([
				'submited' => $arr,
				'expected' => $delta_e,
				'computed' => $res,
				'abs_err' => $abs_err], JSON_PRETTY_PRINT), "\n" ;
				if (++$n_errors == 10)
					break ;
		} else if($i % 1000 === 0){
			echo '.' ;
			fflush(STDOUT);
		}
	}
	fclose($fp) ;
}

chdir(__DIR__);

if (ctype_alpha($argv[1] ?? '-'))
	compare_values($argv[1]);
else
	prepare_values((int)($argv[1] ?? 10000));
